sklearn/linear_model/_glm/glm.py

```diff
@@ -20,13 +24,633 @@
+            #    It may set self.stop = True, e.g. for ill-conditioned systems.
+            self.inner_solve(X=X, y=y, sample_weight=sample_weight)
+            if self.stop:
+                break
+            if self.use_lbfgs_step:
+                self.lbfgs_step(X=X, y=y, sample_weight=sample_weight)
+
+            # 3. Backtracking line search
+            #    This usually sets self.coef_old, self.coef, self.loss_value_old
+            #    self.loss_value, self.gradient_old, self.gradient,
+            #    self.raw_prediction.
+            if not self.use_lbfgs_step:
+                self.line_search(X=X, y=y, sample_weight=sample_weight)
+
+            # 4. Check convergence
+            #    Sets self.converged.
+            self.check_convergence(X=X, y=y, sample_weight=sample_weight)
+
+            # 5. Next iteration
+            self.iteration += 1
+
+        if not self.converged:
+            warnings.warn(
+                "Newton solver did not converge after"
+                f" {self.iteration - 1} iterations.",
+                ConvergenceWarning,
+            )
+
+        self.iteration -= 1
+        self.finalize(X=X, y=y, sample_weight=sample_weight)
+        return self.coef
+
+
+class BaseCholeskyNewtonSolver(NewtonSolver):
+    """Cholesky based Newton solver.
+
+    Inner solver for finding the Newton step H w_newton = -g uses Cholesky based linear
+    solver.
+    """
+
+    def setup(self, X, y, sample_weight):
+        super().setup(X=X, y=y, sample_weight=sample_weight)
+        self.count_singular = 0
+        self.count_hessian_warning = 0
+
+    def inner_solve(self, X, y, sample_weight):
+        if self.hessian_warning:
+            if self.count_bad_hessian == 0:
+                # We only need to throw this warning once.
+                warnings.warn(
+                    f"The inner solver of {self.__class__.__name__} detected a "
+                    " pointwise hessian with many negative values at iteration "
+                    f"#{self.iteration}. It will now try a lbfgs step."
+                    " Note that this warning is only raised once, the problem may,"
+                    " however, occur in several or all iterations. Set verbose >= 1"
+                    " to get more information.\n",
+                    ConvergenceWarning,
+                )
+            self.count_hessian_warning += 1
+            if self.verbose:
+                print(
+                    "  The inner solver detected a pointwise hessian with many "
+                    "negative values and resorts to a lbfgs step."
+                )
+            self.use_lbfgs_step = True
+            return
+
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("error", scipy.linalg.LinAlgWarning)
+                self.coef_newton = scipy.linalg.solve(
+                    self.hessian, -self.gradient, check_finite=False, assume_a="sym"
+                )
```

Comment:
Thanks for trying out. I guess symmetric matrices enables more efficient routines.

Reply:
To me, those results also look rather counter-intuitive: I would have expected `posv` to be more efficient. :thinking:

Reply:
This is interesting. If I remember correctly, Nocedal and Wright use the LDL^T decomposition and call it a Cholesky variant. So I think we can keep the `"newton-cholesky"` name because the LDL^T decomposition should be computationally similar to the traditional Cholesky decomposition.

Reply:
> @jjerphan do you have the same plot in terms of n_iter?

Here it is:

![n_iter](https://user-images.githubusercontent.com/13029839/175002767-ac1c0396-cbf4-4c73-82e4-18c36753036d.svg)

Reply:
So in terms if `n_iter_` LDL^T and "true" Cholesky have overlapping curves but the "true" Cholesky is significantly slower from a Wall-time point of view.

Reply:
LDL' is also called square root free choleksy. Computationally, the assumption of a symmetric matrix is stronger than a positive definiteness: you only need half of the entries. So the results are not entirely surprising - after this second thought.